### ISSUE TITLE
Convenient syntax sugar for coinductive types that could also possibly fix problems with positive coinductives

### ISSUE THREAD

**Issue by wkolowski** [coq/coq#14020]
**TL;DR** **Make the syntax of positive coinductive types into a notation for negative coinductive types with a single constructor.**

Let's say we want to implement some `CoList`s (i.e. coinductive lists, i.e. something like lists, but could possibly be infinite). We know Coq's ordinary equality doesn't play well with coinductives, so we will also implement a custom, well-behaved equality ("bisimilarity").

```coq
Module Negative.

CoInductive CoList (A : Type) : Type :=
{
    uncons : option (A * CoList A);
}.

Arguments uncons {A} _.

Definition Nil {A : Type} : CoList A :=
{|
    uncons := None;
|}.

Definition Cons {A : Type} (h : A) (t : CoList A) : CoList A :=
{|
    uncons := Some (h, t)
|}.

CoInductive sim {A : Type} (l1 l2 : CoList A) : Prop :=
{
    sim' :
      (uncons l1 = None /\ uncons l2 = None)
        \/
      (exists (h1 h2 : A) (t1 t2 : CoList A),
        uncons l1 = Some (h1, t1) /\
        uncons l2 = Some (h2, t2) /\
          h1 = h2 /\ sim t1 t2)
}.

End Negative.
```

With the above code we have fully achieved our goals, but as soon as we try to prove anything, we might find it a little inconvenient. The inconveniences will manifest mostly when we try to `destruct` a `CoList` or a proof of `sim`.

`CoList`s aren't that bad, but `sim` is a mess made of nested disjunctions, conjunctions and existentials. To tear it down we will have to either:
- Spam `destruct`s. This is what novices do when I try to teach this to them, but it isn't very efficient and is probably frustrating.
- Use `decompose` followed with `clear`, like `destruct H; decompose [ex and or] sim'0; clear sim'0`. This is a reasonable solution, but it relies on automatic name generation which is considered harmful for the purposes of proof engineering.
- Use `destruct` with a complex pattern like `destruct H as [[[H1 H2] | (h1 & h2 & t1 & t2 & H1 & H2 & H3 & H4)]]`. This is pretty compact and gets rid of the naming problem, but such patterns are extremely hard to get right off the top of your head, so they come into existence mostly by copy-pasting from previous proofs.
- Write a custom tactic that automates the above. This would yield a pretty nice solution I think (I didn't try that), but it requires knowledge of `Ltac`, which puts it out of reach of mere mortals. The tactic itself would probably also turn out to be quite boilerplate-y.

This would get even more annoying if we were dealing with the analogue of `sim` for something more complicated than `CoList`s, like coinductive trees with 5 different kinds of branching.

But there's a neat trick that can help us.

```coq
Module MuLike.

Inductive CoListF (F : Type -> Type) (A : Type) : Type :=
    | NilF  : CoListF F A
    | ConsF : forall (h : A) (t : F A), CoListF F A.

Arguments NilF {F A}.
Arguments ConsF {F A} _ _.

CoInductive CoList (A : Type) : Type :=
{
    uncons : CoListF CoList A;
}.

Arguments uncons {A} _.

Definition Nil {A : Type} : CoList A :=
{|
    uncons := NilF;
|}.

Definition Cons {A : Type} (h : A) (t : CoList A) : CoList A :=
{|
    uncons := ConsF h t;
|}.

Inductive simF
  (F : forall A : Type, CoList A -> CoList A -> Prop)
  {A : Type} (l1 l2 : CoList A)  : Prop :=
    | Nils  :
        forall (H1 : uncons l1 = NilF) (H2 : uncons l2 = NilF), simF F l1 l2
    | Conss :
        forall
          (h1 h2 : A) (t1 t2 : CoList A)
          (H1 : uncons l1 = ConsF h1 t1)
          (H2 : uncons l2 = ConsF h2 t2)
          (heads : h1 = h2) (tails : F A t1 t2),
            simF F l1 l2.

CoInductive sim {A : Type} (l1 l2 : CoList A) : Prop :=
{
    sim' : simF (@sim) l1 l2
}.

End MuLike.
```

This time we define inductive "templates" (which probably are the polynomial functors whose greatest fixpoints are `CoList` and `sim`, or something like that) and then tie the knot using a simple, single field `CoInductive`.

It's a pretty good solution of our problem - to `destruct` a proof of `sim`, we can now simply write `destruct H as [[]]` and get very good default names for the components, while avoiding spamming `destruct`s, `decompose`, complex patterns and Ltac. But it's not a perfect solution - the cost is boilerplate and, most likely, increased cognitive load (I bet this style of definition would be confusing for beginners). Also note the annoying layer of indirection, which causes us to write `destruct [[]]` instead of just `destruct []`.

Is there a way to improve upon this? Well, Coq has a nice syntax that could be of some use...

```coq
Module Positive.

CoInductive CoList (A : Type) : Type :=
    | Nil  : CoList A
    | Cons : forall (h : A) (t : CoList A), CoList A.

Arguments Nil {A}.
Arguments Cons {A} _ _.

Definition uncons {A : Type} (l : CoList A) : option (A * CoList A) :=
match l with
    | Nil      => None
    | Cons h t => Some (h, t)
end.

CoInductive sim {A : Type} : CoList A -> CoList A -> Prop :=
    | Nils  : sim Nil Nil
    | Conss :
        forall
          (h1 h2 : A) (t1 t2 : CoList A)
          (heads : h1 = h2) (tails : sim t1 t2),
            sim (Cons h1 t1) (Cons h2 t2).

End Positive.
```

... but sadly, this syntax is already taken by positive coinductive types, which are broken #7536.

So I propose to remove positive coinductive types and repurpose this existing syntax as syntax sugar for the trick presented above. If inductive records can be realized as syntax sugar for single-constructor inductives, we should have some syntax sugar for coinductive sums (as single-constructor coinductives) too!

**Comment by SkySkimmer** [comment]
Or just go with https://github.com/coq/coq/pull/10764

**Comment by ppedrot** [comment]
I agree that it would be neat to have some syntactic sugar to define sum types inside negative coinductive records. As you observe, it is pretty heavy currently. But I think it's a very bad idea to reuse the positive coinductive syntax, because it is going to hide critical details to the user. If I have learned something from primitive projections, it is that pretending that two distinct kernel terms have the same user-facing representation without a way to discriminate them can easily drive anybody crazy.
